pyonfx/timestamps.py

```diff
@@ -0,0 +1,552 @@
+                highest_timestamp = timestamp
+            if lowest_timestamp is None or lowest_timestamp > timestamp:
+                lowest_timestamp = timestamp
+
+            rounded_timestamp = rounding_method(timestamp)
+
+            if version == 2 and rounded_timestamp < previous_timestamp:
+                raise ValueError(
+                    f"The timestamps file contain timestamps NOT in ascending order."
+                )
+
+            previous_timestamp = rounded_timestamp
+            timestamps.append(rounded_timestamp)
+
+        if not len(timestamps):
+            raise ValueError(f"The timestamps file is empty.")
+
+        if version == 4:
+            timestamps.sort()
+
+        duration = highest_timestamp - lowest_timestamp
+        if duration:
+            fpms = Fraction(len(timestamps) - 1) / (
+                highest_timestamp - lowest_timestamp
+            )
+        else:
+            fpms = 0
+        return timestamps, highest_timestamp, fpms
+
+
+class Timestamps:
+    """Timestamps object contains informations about the timestamps of an video.
+
+    Video player have 2 methods to deal with timestamps. Some floor them and other round them.
+    This can lead to difference when displaying the subtitle.
+        Ex:
+            Player - Method - proof
+            mpv    - round  - https://github.com/mpv-player/mpv/blob/7480efa62c0a2a1779b4fdaa804a6512aa488400/sub/sd_ass.c#L499
+            FFmpeg - floor  - https://github.com/FFmpeg/FFmpeg/blob/fd1712b6fb8b7acc04ccaa7c02b9a5c9f233cfb3/libavfilter/vf_subtitles.c#L195
+            VLC    - floor  - https://github.com/videolan/vlc/blob/f7bb59d9f51cc10b25ff86d34a3eff744e60c46e/include/vlc_tick.h#L118-L130
+            MPC-HC - floor  - https://github.com/clsid2/mpc-hc/blob/0994fd605a9fb4d15806d0efdd6399ba1bc5f984/src/Subtitles/LibassContext.cpp#L843
+    Important note:
+        Matroska (.mkv) file are an exception !!!
+        If you want to be compatible with mkv, use RoundingMethod.ROUND.
+        By default, they only have a precision to milliseconds instead of nanoseconds like most format.
+            For more detail see:
+                1- https://mkvtoolnix.download/doc/mkvmerge.html#mkvmerge.description.timestamp_scale
+                2- https://matroska.org/technical/notes.html#timestampscale-rounding
+
+    Parameters:
+        rounding_method (RoundingMethod): A rounding method. See the comment above about floor vs round.
+            99% of the time, you want to use RoundingMethod.ROUND.
+        timestamps (List[int], optional): A list of [timestamps](https://en.wikipedia.org/wiki/Timestamp) in milliseconds encoded as integers.
+                                It represent each frame [presentation timestamp (PTS)](https://en.wikipedia.org/wiki/Presentation_timestamp)
+        normalize (bool, optional): If True, it will shift the timestamps to make them start from 0. If false, the option does nothing.
```

Comment:
At 99% of the time, the user want the normalize timestamps.
With you idea, this means the user would need to do 2 calls VS 1.
Ex:
Before
```py
timestamps = Timestamps.from_video_file(video_path) # by default, it normalize it
```

After
```py
timestamps = Timestamps.from_video_file(video_path)
timestamps.normalize()
```

Do you still want me to add your idea?

Reply:
Mhhh... I think I am not getting the idea of normalizing then, mostly for VFR videos: if the frame-rate is different in different parts of the video, doesn't normalizing the timestamps yield a bad result? As we will misalign the timestamps with the video.

Reply:
> Mhhh... I think I am not getting the idea of normalizing then, mostly for VFR videos:

But I am not sure if you totally understand what the method does.
It only shift the timestamps so the first one start at 0.

> doesn't normalizing the timestamps yield a bad result?

Yes and no. This is a very very tricky subject that depend on the video player (i am pretty sure that they don't act all the same, but I don't have any proof of that)

Honestly, I don't have the level to explain properly why it is necessary. I just know that sometimes, the video player will normalized them depending on the timestamps of the audio and/or chapter.

Just for you to know, [Aegisub](https://github.com/arch1t3cht/Aegisub/blob/752a14cc25f361b2c5265d93561fe997e78d576d/libaegisub/common/vfr.cpp#L48-L53) always normalize the video timestamps.

99% of the video are already normalize.

Reply:
Mhhhh... Do you know what MPV does in this respect?

Reply:
From what I know, MPV will take the first timestamps of the video or audio or chapter and normalize from that.

**Ex 1 - All timestamps equals:**
video first timestamps = 83 ms
audio first timestamps = 83 ms
chapter first timestamps = 83 ms
Then, it will set the video first timestamps to 0.

**Ex 2 - Different timestamps:**
video first timestamps = 83 ms
audio first timestamps = 83 ms
chapter first timestamps = 0 ms
Then, it will set the video first timestamps to 83

Reply:
Ok. Your solution seems legit then.
Thank you for the explanation.